File: lang/haskell/learnyouhaskell/ch12/Bird_Monad.hs
type Birds = Int
type Pole = (Birds, Birds)

check :: Pole -> Maybe Pole
check (left, right)
        | abs (left - right) < 4 = Just (left, right)
        | otherwise = Nothing

landLeft :: Birds -> Pole -> Maybe Pole
landLeft n (left, right) = check (left + n, right)

landRight :: Birds -> Pole -> Maybe Pole
landRight n (left, right) = check (left, right + n)

banana _ = Nothing

-- Maybe is a Monad, which handle failure
ret0 = return (0, 0) >>= landLeft 1 >>= landRight 1 >>= landLeft 2
ret1 = return (0, 0) >>= landLeft 5 >>= landRight 1 >>= landLeft (-2)
ret2 = return (0, 0) >>= landLeft 1 >>= banana >>= landLeft 2
{-
    (>>) :: (Monad m) => m a -> m b -> m b
    m >> n = m >>= \_ -> n
-}
-- use >>, we can directly make failure
ret3 = return (0, 0) >>= landLeft 1 >> Nothing >>= landLeft 2

main = do print ret0
          print ret1
          print ret2
          print ret3

File: lang/c/HaskellInC/Bird_Monad.c
#include <stdio.h>
#include <stdlib.h>
typedef int Birds;

typedef enum {
    Just = 0,
    Nothing = 1,
} MaybeT;

typedef struct {
    Birds left;
    Birds right;
} Pole;

typedef struct {
    MaybeT maybe;
    Pole pole;
} MaybePole;

MaybePole JustPole(Pole pole) {
    MaybePole r = {.maybe = Just, .pole = pole};
    return r;
}

MaybePole NothingPole() {
    MaybePole r = {.maybe = Nothing};
    return r;
}

/* basic function for Maybe monad */
MaybePole returnMaybe(Pole pole) {
    return JustPole(pole);
}

MaybePole bind(MaybePole mpole, MaybePole(*func)(Pole pole)) {
    if (mpole.maybe == Nothing) {
        return NothingPole();
    }
    return func(mpole.pole);
}

void dispMaybePole(MaybePole mpole) {
    if (mpole.maybe == Nothing) {
        printf("Nothing\n");
    } else {
        printf("Just (%d, %d)\n", mpole.pole.left, mpole.pole.right);
    }
}

MaybePole check(Pole pole) {
    if (abs(pole.right - pole.right) < 4) {
        return JustPole(pole);
    }
    return NothingPole();
}

MaybePole landLeft(Birds n, Pole pole) {
    Pole r = {.left = pole.left + n, .right = pole.right};
    return check(r);
}

MaybePole landRight(Birds n, Pole pole) {
    Pole r = {.left = pole.left, .right = pole.right + n};
    return check(r);
}

// sorry, we don't have currying in C.
MaybePole landLeft1(Pole pole) {
    return landLeft(1, pole);
}

MaybePole landRight1(Pole pole) {
    return landRight(1, pole);
}

MaybePole landLeft2(Pole pole) {
    return landLeft(2, pole);
}

MaybePole banana(Pole pole) {
    return NothingPole();
}

void test0() {
    Pole initial = {.left = 0, .right = 0};
    MaybePole ret = bind(bind(bind(returnMaybe(initial),
                                   landLeft1),
                              landRight1),
                         landLeft2);
    dispMaybePole(ret);
}
void test2() {
    Pole initial = {.left = 0, .right = 0};
    MaybePole ret = bind(bind(bind(returnMaybe(initial),
                                   landLeft1),
                              banana),
                         landLeft2);
    dispMaybePole(ret);
}
int main() {
    test0();
    test2();
    return 0;
}
